libhsail-rt/rt/ftz.c

```diff
@@ -0,0 +1,176 @@
+/* ftz.c -- Builtins for FTZ routines.
+
+   Copyright (C) 2015-2017 Free Software Foundation, Inc.
+   Contributed by Henry Linjamaki <henry.linjamaki@parmance.com>
+   for General Processor Tech.
+
+   Permission is hereby granted, free of charge, to any person obtaining a
+   copy of this software and associated documentation files
+   (the "Software"), to deal in the Software without restriction, including
+   without limitation the rights to use, copy, modify, merge, publish,
+   distribute, sublicense, and/or sell copies of the Software, and to
+   permit persons to whom the Software is furnished to do so, subject to
+   the following conditions:
+
+   The above copyright notice and this permission notice shall be included
+   in all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+   OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+   MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+   IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+   DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+   OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+   USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#include "ftz.h"
+
+#include <stdint.h>
+#include <float.h>
+#include <math.h>
+#if defined(___x86_64_) || defined(__SSE__)
+#  include "xmmintrin.h"
+#  include "pmmintrin.h"
+#endif
+
+/* Sets flags to enable flush to zero if applicable on the device.
+   Returns original flags prior to the call. */
+
+#define INTEL_FTZ_POS 0u
+#define INTEL_DAZ_POS 1u
+
+uint64_t
+__hsail_enable_ftz ()
+{
+#if defined(___x86_64_) || defined(__SSE__)
+  uint32_t ftz_flag = _MM_GET_FLUSH_ZERO_MODE ();
+  uint32_t daz_flag = _MM_GET_DENORMALS_ZERO_MODE();
+
+  _MM_SET_FLUSH_ZERO_MODE (_MM_FLUSH_ZERO_ON);
+  _MM_SET_DENORMALS_ZERO_MODE(_MM_DENORMALS_ZERO_ON);
+
+  uint64_t saved_flags
+    = (!!ftz_flag << INTEL_FTZ_POS) | (!!daz_flag << INTEL_DAZ_POS);
+  return saved_flags;
+#else
+#  error "Unsupported target."
```

Comment:
Could it be possible to avoid this error and fall back to the awfully slow mode of always calling the builtins in case CPU mode flag toggling is not implemented for the target?